pallets/bridge_transfer/src/lib.rs

```diff
@@ -179,122 +265,93 @@ pub mod pallet {
-					), // roundId
-					u32::from_be_bytes(
-						<[u8; 4]>::try_from(&metadata[5..9])
-							.map_err(|_| Error::<T>::InvalidCommand)?,
-					), // totalCount
-					u32::from_be_bytes(
-						<[u8; 4]>::try_from(&metadata[9..])
-							.map_err(|_| Error::<T>::InvalidCommand)?,
-					), // winnerCount
-				));
-			} else if op == 1 {
-				ensure!(metadata.len() > 13, Error::<T>::InvalidCommand);
-
-				let address_len: usize = u32::from_be_bytes(
-					<[u8; 4]>::try_from(&metadata[9..13])
-						.map_err(|_| Error::<T>::InvalidCommand)?,
-				)
-				.saturated_into();
-				ensure!(
-					metadata.len() == (13 + address_len),
-					Error::<T>::InvalidCommand
-				);
-
-				Self::push_command(LotteryCommand::open_box(
-					u32::from_be_bytes(
-						<[u8; 4]>::try_from(&metadata[1..5])
-							.map_err(|_| Error::<T>::InvalidCommand)?,
-					), // roundId
-					u32::from_be_bytes(
-						<[u8; 4]>::try_from(&metadata[5..9])
-							.map_err(|_| Error::<T>::InvalidCommand)?,
-					), // tokenId
-					metadata[13..].to_vec(), // btcAddress
-				));
+			if _rid == T::ReserveTokenId::get() {
+				<T as Config>::Currency::transfer(
+					&source,
+					&to,
+					amount,
+					ExistenceRequirement::AllowDeath,
+				)?;
 			} else {
-				fail!(Error::<T>::InvalidCommand);
+				// if recipient isn't holding account, make sure holding accouint balance to cover transfer amount
+				if to != source {
+					ensure!(
+						Self::asset_balance(&_rid, &source) >= amount,
+						Error::<T>::InsufficientBalance
+					);
+				}
+				Self::do_asset_deposit(&_rid, &to, amount);
 			}
 
 			Ok(())
 		}
-
-		#[pallet::weight(0)]
-		pub fn force_lottery_new_round(
-			origin: OriginFor<T>,
-			round_id: u32,
-			total_count: u32,
-			winner_count: u32,
-		) -> DispatchResult {
-			ensure_root(origin)?;
-			Self::push_command(LotteryCommand::new_round(
-				round_id,
-				total_count,
-				winner_count,
-			));
-			Ok(())
-		}
-
-		#[pallet::weight(0)]
-		pub fn force_lottery_open_box(
-			origin: OriginFor<T>,
-			round_id: u32,
-			token_id: u32,
-			btc_address: Vec<u8>,
-		) -> DispatchResult {
-			ensure_root(origin)?;
-			Self::push_command(LotteryCommand::open_box(round_id, token_id, btc_address));
-			Ok(())
-		}
 	}
 
 	impl<T: Config> MessageOriginInfo for Pallet<T> {
 		type Config = T;
 	}
 
 	impl<T: Config> Pallet<T> {
-		pub fn lottery_output(payload: &Lottery, dest_id: bridge::BridgeChainId) -> DispatchResult {
-			ensure!(
-				<bridge::Pallet<T>>::chain_whitelisted(dest_id),
-				Error::<T>::InvalidTransfer
-			);
-			let metadata: Vec<u8> = payload.encode();
-			<bridge::Pallet<T>>::transfer_generic(dest_id, T::BridgeLotteryId::get(), metadata)
+		pub fn asset_balance(asset: &bridge::ResourceId, who: &T::AccountId) -> BalanceOf<T> {
+			BridgeBalances::<T>::get(asset, who).unwrap_or(Zero::zero())
+		}
+
+		pub fn calculate_fee(dest_id: bridge::BridgeChainId, amount: BalanceOf<T>) -> BalanceOf<T> {
```

Comment:
This function takes the amount of the asset as the input, and output the corresponding fee in PHA. But we didn't consider any currency exchange rate here. It's not an easy thing unless we got an price oracle. However let's leave a TODO here to address the problem.